flink-python/pyflink/fn_execution/operations.py

```diff
@@ -0,0 +1,261 @@
+        return self.scalar_function.eval(*args)
+
+
+def create_scalar_function_invoker(scalar_function_proto):
+    """
+    Creates :class:`ScalarFunctionInvoker` from the proto representation of a
+    :class:`ScalarFunction`.
+
+    :param scalar_function_proto: the proto representation of the Python :class:`ScalarFunction`
+    :return: :class:`ScalarFunctionInvoker`.
+    """
+    import cloudpickle
+    scalar_function = cloudpickle.loads(scalar_function_proto.payload)
+    return ScalarFunctionInvoker(scalar_function, scalar_function_proto.inputs)
+
+
+class ScalarFunctionRunner(object):
+    """
+    The runner which is responsible for executing the scalar functions and send the
+    execution results back to the remote Java operator.
+
+    :param udfs_proto: protocol representation for the scalar functions to execute
+    :param main_receivers: Receiver objects which is responsible for sending the execution
+                           results back the the remote Java operator
+    """
+
+    def __init__(self, udfs_proto, main_receivers):
+        self.scalar_function_invokers = [create_scalar_function_invoker(f) for f in
+                                         udfs_proto]
+        from apache_beam.runners.common import _OutputProcessor
+        self.output_processor = _OutputProcessor(
+            window_fn=None,
+            main_receivers=main_receivers,
+            tagged_receivers=None,
+            per_element_output_counter=None)
+
+    def open(self):
+        for invoker in self.scalar_function_invokers:
+            invoker.invoke_open()
+
+    def close(self):
+        for invoker in self.scalar_function_invokers:
+            invoker.invoke_close()
+
+    def start(self):
+        pass
+
+    def finish(self):
+        pass
+
+    def process(self, windowed_value):
+        results = [invoker.invoke_eval(windowed_value.value) for invoker in
+                   self.scalar_function_invokers]
+        from pyflink.table import Row
+        result = Row(*results)
+        # send the execution results back
+        self.output_processor.process_outputs(windowed_value, [result])
+
+
+class ScalarFunctionOperation(Operation):
+    """
+    An operation that will execute ScalarFunctions for each input element.
+    """
+
+    def __init__(self, name, spec, counter_factory, sampler, consumers):
+        super(ScalarFunctionOperation, self).__init__(name, spec, counter_factory, sampler)
+        for tag, op_consumers in consumers.items():
+            for consumer in op_consumers:
+                self.add_receiver(consumer, 0)
+
+    def setup(self):
+        with self.scoped_start_state:
+            super(ScalarFunctionOperation, self).setup()
+
+            self.scalar_function_runner = ScalarFunctionRunner(
+                self.spec.serialized_fn,
+                main_receivers=self.receivers[0])
+
+            self.scalar_function_runner.open()
```

Comment:
setup() will be called at the beginning of each bundle while the open() method can only be called once.